src/value/number.rs

```diff
@@ -0,0 +1,410 @@
+enum N {
+    PosInt(u64),
+    /// Always less than zero.
+    NegInt(i64),
+    /// Always finite.
+    Float(f64),
+}
+
+
+impl Number {
+    /// Returns true if the `Number` is an integer between `i64::MIN` and
+    /// `i64::MAX`.
+    ///
+    /// For any Number on which `is_i64` returns true, `as_i64` is guaranteed to
+    /// return the integer value.
+    ///
+    /// ```rust
+    /// # #[macro_use]
+    /// # extern crate serde_yaml;
+    /// #
+    /// # use std::i64;
+    /// #
+    /// # fn yaml(i: &str) -> serde_yaml::Value { serde_yaml::from_str(i).unwrap() }
+    /// # fn main() {
+    /// let big = i64::MAX as u64 + 10;
+    /// let v = yaml(r#"
+    /// a: 64
+    /// b: 9223372036854775817
+    /// c: 256.0
+    /// "#);
+    ///
+    /// assert!(v["a"].is_i64());
+    ///
+    /// // Greater than i64::MAX.
+    /// assert!(!v["b"].is_i64());
+    ///
+    /// // Numbers with a decimal point are not considered integers.
+    /// assert!(!v["c"].is_i64());
+    /// # }
+    /// ```
+    #[inline]
+    pub fn is_i64(&self) -> bool {
+        match self.n {
+            N::PosInt(v) => v <= i64::MAX as u64,
+            N::NegInt(_) => true,
+            N::Float(_) => false,
+        }
+    }
+
+    /// Returns true if the `Number` is an integer between zero and `u64::MAX`.
+    ///
+    /// For any Number on which `is_u64` returns true, `as_u64` is guaranteed to
+    /// return the integer value.
+    ///
+    /// ```rust
+    /// # #[macro_use]
+    /// # extern crate serde_yaml;
+    /// #
+    /// # fn yaml(i: &str) -> serde_yaml::Value { serde_yaml::from_str(i).unwrap() }
+    /// # fn main() {
+    /// let v = yaml(r#"
+    /// a: 64
+    /// b: -64
+    /// c: 256.0
+    /// "#);
+    ///
+    /// assert!(v["a"].is_u64());
+    ///
+    /// // Negative integer.
+    /// assert!(!v["b"].is_u64());
+    ///
+    /// // Numbers with a decimal point are not considered integers.
+    /// assert!(!v["c"].is_u64());
+    /// # }
+    /// ```
+    #[inline]
+    pub fn is_u64(&self) -> bool {
+        match self.n {
+            N::PosInt(_) => true,
+            N::NegInt(_) | N::Float(_) => false,
+        }
+    }
+
+    /// Returns true if the `Number` can be represented by f64.
+    ///
+    /// For any Number on which `is_f64` returns true, `as_f64` is guaranteed to
+    /// return the floating point value.
+    ///
+    /// Currently this function returns true if and only if both `is_i64` and
+    /// `is_u64` return false but this is not a guarantee in the future.
+    ///
+    /// ```rust
+    /// # #[macro_use]
+    /// # extern crate serde_yaml;
+    /// #
+    /// # fn yaml(i: &str) -> serde_yaml::Value { serde_yaml::from_str(i).unwrap() }
+    /// # fn main() {
+    /// let v = yaml(r#"
+    /// ---
+    /// a: 256.0
+    /// b: 64
+    /// c: -64
+    /// "#);
+    ///
+    /// assert!(v["a"].is_f64());
+    ///
+    /// // Integers.
+    /// assert!(!v["b"].is_f64());
+    /// assert!(!v["c"].is_f64());
```

Comment:
These two asserts are failing and I can't reason out how `!v["b"].is_f64()` could be returning `true`.